engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -63,52 +92,277 @@ GfxDefaultDriver::GfxDefaultDriver(uint16 screenWidth, uint16 screenHeight) : Gf
+	// updated on the next OSysten::updateScreen() call. This is not what we need here, so I rather use
+	// the extra bitmap buffer (which is required for rgb rendering anyway).
+	memcpy(dest, _currentBitmap, size);
 }
 
-SCI0_DOSPreVGADriver::~SCI0_DOSPreVGADriver() {
-	delete[] _compositeBuffer;
+void GfxDefaultDriver::copyCurrentPalette(byte *dest, int start, int num) const {
+	GFXDRV_ASSERT_READY;
+
+	if (_pixelSize == 1) {
+		GfxDriver::copyCurrentPalette(dest, start, num);
+		return;
+	}
+
+	assert(dest);
+	assert(_currentPalette);
+	assert(start + num <= 256);
+	memcpy(dest + start * 3, _currentPalette + start * 3, num * 3);
 }
 
-bool SCI0_DOSPreVGADriver::checkDriver(const char *const *driverNames, int listSize) {
-	Common::String missing;
-	while (listSize-- && *driverNames) {
-		if (Common::File::exists(*driverNames))
-			return true;
-		if (!missing.empty())
-			missing += " or ";
-		missing += "'" + Common::String(*driverNames) + "'";
-		++driverNames;
+template <typename T> void updateRGBPalette(byte *dest, const byte *src, uint start, uint num, Graphics::PixelFormat &f) {
+	T *dst = &reinterpret_cast<T*>(dest)[start];
+	for (uint i = 0; i < num; ++i) {
+		*dst++ = f.RGBToColor(src[0], src[1], src[2]);
+		src += 3;
 	}
-	warning("Driver file %s not found. Starting game in EGA mode", missing.c_str());
-	return false;
+}
+
+void GfxDefaultDriver::updatePalette(const byte *colors, uint start, uint num, bool skipRGBPalette) {
+	memcpy(_currentPalette + start * 3, colors, num * 3);
+	// If palette mods are on we don't need to update the internal palette,
+	// since the colors have to be generated on the fly anyway.
+	if (skipRGBPalette)
+		return;
+	if (_pixelSize == 4)
+		updateRGBPalette<uint32>(_internalPalette, colors, start, num, _format);
+	else if (_pixelSize == 2)
+		updateRGBPalette<uint16>(_internalPalette, colors, start, num, _format);
+	else
+		error("GfxDefaultDriver::updatePalette(): Unsupported pixel size %d", _pixelSize);
+}
+
+template <typename T> void render(byte *dst, const byte *src, int pitch, int w, int h, const byte *pal) {
+	T *d = reinterpret_cast<T*>(dst);
+	const T *p = reinterpret_cast<const T*>(pal);
+	const byte *s = src;
+	pitch -= w;
+
+	while (h--) {
+		for (int i = 0; i < w; ++i)
+			*d++ = p[*s++];
+		s += pitch;
+	}
+}
+
+#define applyMod(a, b) MIN<uint>(a * (128 + b) / 128, 255)
+template <typename T> void renderMod(byte *dst, const byte *src, int pitch, int w, int h, const byte *pal, Graphics::PixelFormat &f, const PaletteMod *mods, const byte *modMapping) {
+	T *d = reinterpret_cast<T*>(dst);
+	const byte *s1 = src;
+	const byte *s2 = modMapping;
+	pitch -= w;
+
+	while (h--) {
+		for (int i = 0; i < w; ++i) {
+			const byte *col = &pal[*s1++ * 3];
+			byte m = *s2++;
+			if (m)
+				*d++ = f.RGBToColor(applyMod(col[0], mods[m].r), applyMod(col[1], mods[m].g), applyMod(col[2], mods[m].b));
+			else
+				*d++ = f.RGBToColor(col[0], col[1], col[2]);
```

Comment:
Could this get the value from `_internalPalette` if a mod colour isn't being applied?

Reply:
I think that would make sense. When I tested SQ3 and LSL3, my impression was that most pixels won't be be affected by the customizations. These seem to target mostly the actors. So it would make sense to still use the pre-made rgb colors here. I'll make that change.

Reply:
Hmm, just tested again, these mods do hit much more pixels than I thought, but the change still makes sense.